modules/heat_conduction/src/userobjects/ViewFactorRayStudy.C

```diff
@@ -0,0 +1,554 @@
+                                        const THREAD_ID tid)
+{
+  mooseAssert(currentlyPropagating(), "Can only be called during Ray tracing");
+  mooseAssert(_threaded_vf_info[tid].count(from_id),
+              "Threaded view factor info does not have from boundary");
+  mooseAssert(_threaded_vf_info[tid][from_id].count(to_id),
+              "Threaded view factor info does not have from -> to boundary");
+
+  _threaded_vf_info[tid][from_id][to_id] += value;
+}
+
+Real
+ViewFactorRayStudy::viewFactorInfo(const BoundaryID from_id, const BoundaryID to_id) const
+{
+  auto it = _vf_info.find(from_id);
+  if (it == _vf_info.end())
+    mooseError("From boundary id ", from_id, " not in view factor map.");
+
+  auto itt = it->second.find(to_id);
+  if (itt == it->second.end())
+    mooseError("From boundary id ", from_id, " to boundary_id ", to_id, " not in view factor map.");
+  return itt->second;
+}
+
+void
+ViewFactorRayStudy::generateStartElems()
+{
+  const auto & points = _fe_face->get_xyz();
+  const auto & weights = _fe_face->get_JxW();
+
+  // Clear before filling
+  _start_elems.clear();
+
+  // Starting elements we have that are on the wrong side of an internal boundary
+  std::unordered_map<processor_id_type, std::vector<StartElem>> send_start_map;
+
+  // Get all possible points on the user defined boundaries on this proc
+  for (const BndElement * belem : *_mesh.getBoundaryElementRange())
+  {
+    const Elem * elem = belem->_elem;
+    const auto side = belem->_side;
+    const auto bnd_id = belem->_bnd_id;
+
+    // Skip if we don't own you
+    if (elem->processor_id() != _pid)
+      continue;
+
+    // Skip if the boundary id isn't one we're looking for
+    if (!_bnd_ids.count(bnd_id))
+      continue;
+
+    // Sanity check on QGRID not working on some types
+    if (_q_face->type() == QGRID && elem->type() == TET4)
+      mooseError(
+          "Cannot use GRID quadrature type with tetrahedral elements in ViewFactorRayStudy '",
+          _name,
+          "'");
+
+    // The elem/side that we will actually start the trace from
+    // (this may change on internal sidesets)
+    const Elem * start_elem = elem;
+    auto start_side = side;
+
+    // Reinit this face for points
+    _fe_face->reinit(elem, side);
+
+    // See if this boundary is internal
+    const Elem * neighbor = elem->neighbor_ptr(side);
+    if (neighbor)
+    {
+      if (!neighbor->active())
+        mooseError(type(), " does not work with adaptivity");
+
+      // With the positive convention, the Rays that we want to spawn from internal boundaries
+      // have positive dot products with the outward normal on the side. The ray-tracer requires
+      // that we provide an element incoming side that is actually incoming (the dot product with
+      // the direction and the normal is negative). Therefore, switch the physical trace to start
+      // from the other element and the corresponding side
+      if (_internal_convention == 0)
```

Comment:
*_mesh.getBoundaryElementRange() is consistently giving us elem with negative dot products then?

Reply:
The range is consistent in that it returns the sides that are on boundaries, as it should. It just so happens that we could also spawn rays in the other direction away from the normal, which ray tracing doesn't like, which is where we switch the ray to start from the other elem.